proxy/proxy/client.go

```diff
@@ -158,24 +164,28 @@ func (c *Client) refreshCfg(instance string) (addr string, cfg *tls.Config, err
 		throttle = DefaultRefreshCfgThrottle
 	}
 
-	if old := c.cfgCache[instance]; time.Since(old.lastRefreshed) < throttle {
+	c.cacheL.Lock()
+	if c.cfgCache == nil {
+		c.cfgCache = make(map[string]cacheEntry)
+	}
+	old, oldok := c.cfgCache[instance]
+	c.cacheL.Unlock()
+
+	if oldok && time.Since(old.lastRefreshed) < throttle {
 		logging.Errorf("Throttling refreshCfg(%s): it was only called %v ago", instance, time.Since(old.lastRefreshed))
 		// Refresh was called too recently, just reuse the result.
 		return old.addr, old.cfg, old.err
 	}
 
-	if c.cfgCache == nil {
-		c.cfgCache = make(map[string]cacheEntry)
-	}
-
 	defer func() {
+		c.cacheL.Lock()
 		c.cfgCache[instance] = cacheEntry{
```

Comment:
We should enhance the cache update logic a bit. If there is an existing cache entry which is valid (no error, and the cert isn't expired) we should avoid overwriting that cache entry with a new one if we have an error getting a new one. We might as well allow the user to use the other cert for 5 more minutes and the normal "redial when expired" mechanism will work as expected. (we don't need to set off another background goroutine or retry in the background, I think an error here will be rare enough.)

Reply:
Done, I think it's enough to check oldok.